deepspeed/runtime/zero/stage2.py

```diff
@@ -1993,8 +2003,12 @@ def _restore_from_fp32_weights(self, all_state_dict):
             ]
             flat_merged_partitions = self.flatten_dense_tensors_aligned(
                 merged_partitions,
+                self.nccl_start_alignment_factor *
                 dist.get_world_size(group=self.real_dp_process_group[i]))
-            dp_partitions = self.get_data_parallel_partitions(flat_merged_partitions, i)
+            dp_partitions = self.get_data_parallel_partitions(
+                flat_merged_partitions,
+                i,
+                self.nccl_start_alignment_factor)
```

Comment:
Again, since flat_merged_partitions is aligned, we not need the alignment here but we can add an assertion that dp_partitions are all aligned to nccl_start_aligned_factor

Reply:
Same as my comment above.

To be specific, the nccl alignment we are concerned about here is the alignment of the start location of each partition. See the newly added testcases.

Reply:
please see comment above